lisa/_assets/kmodules/lisa/fs.c

```diff
@@ -0,0 +1,755 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+
+#include <linux/fs.h>
+#include <linux/fs_context.h>
+#include <linux/seq_file.h>
+#include <linux/slab.h>
+#include <linux/user_namespace.h>
+#include <linux/percpu-rwsem.h>
+
+#include "features.h"
+#include "configs.h"
+
+static int lisa_fs_create_files(struct dentry *dentry, bool is_top_level, struct lisa_cfg *cfg);
+static struct dentry *
+lisa_fs_create_single(struct dentry *parent, const char *name,
+		      const struct inode_operations *i_ops,
+		      const struct file_operations *f_ops, umode_t mode,
+		      void *data);
+
+#define LISA_FS_SUPER_MAGIC 0xcdb11bc9
+
+/* Protect the interface. */
+static struct mutex interface_lock;
+
+int feature_param_lisa_validate(struct feature_param_entry_value *val)
+{
+	struct feature *feature;
+
+	for_each_feature(feature) {
+		if (!strcmp(feature->name, val->data))
+			return 0;
+	}
+	return -EINVAL;
+}
+
+/* Handle feature names using the (struct feature_param) logic. */
+struct feature_param lisa_features_param = {
+	.name = "lisa_features_param",
+	.mode = FT_PARAM_MODE_SET,
+	.type = FT_PARAM_TYPE_AVAILABLE_FT,
+	.perms = S_IFREG | S_IRUGO | S_IWUGO,
+	.ops = &feature_param_ops_string,
+	.validate = feature_param_lisa_validate,
+	.param_args = HLIST_HEAD_INIT,
+	.global_value = LIST_HEAD_INIT(lisa_features_param.global_value),
+};
+
+static struct inode *lisa_fs_create_inode(struct super_block *sb, int mode)
+{
+	struct inode *inode = new_inode(sb);
+
+	if (inode) {
+		inode->i_ino = get_next_ino();
+		inode->i_mode = mode;
+		inode->i_atime = inode->i_mtime = current_time(inode);
+	}
+
+	return inode;
+}
+
+/*
+ * available_features handlers
+ */
+
+static int lisa_features_available_show(struct seq_file *s, void *data)
+{
+	struct feature *feature;
+
+	for_each_feature(feature)
+		if (!feature->__internal)
+			seq_printf(s, "%s\n", feature->name);
+
+	return 0;
+}
+
+static int lisa_features_available_open(struct inode *inode, struct file *file)
+{
+	return single_open(file, lisa_features_available_show, NULL);
+}
+
+static struct file_operations lisa_available_features_fops = {
+	.owner = THIS_MODULE,
+	.open = lisa_features_available_open,
+	.read = seq_read,
+	.llseek = seq_lseek,
+	.release = single_release,
+};
+
+/*
+ * activate handlers
+ */
+
+static int lisa_activate_show(struct seq_file *s, void *data)
+{
+	struct lisa_cfg *cfg = s->private;
+
+	seq_printf(s, "%d\n", cfg->activated);
+	return 0;
+}
+
+static int update_global_value(struct lisa_cfg *cfg, int new_value)
```

Comment:
Why  not placing it in config,c where it feels like a more obvious choice ?

Reply:
No specific reason, I will move it